### ISSUE TITLE
softmax_cross_entropy does not check for input validity when run on gpu (but does so on cpu)

### ISSUE THREAD

**Issue by fabiencro** [pfnet/chainer#971]
Hi,

I am not sure if it is a bug or intended for performance reason, but I was recently surprised by the following behaviour:

if the "t" argument to softmax_cross_entropy is too large (with respect to the "x" argument), it raises an exception on cpu but not on gpu. Below is a sample code demonstrating that behaviour. If you run it, you can see that softmax_cross_entropy with an invalid t still return a value when run on gpu (but raises an exception when run on cpu)

If this is done for performance reasons, it might be good to mention in the doc that gpu mode do less checks than cpu mode.

That being said, I like a lot your library, so thank you for it :-)

```
import chainer
from chainer import Variable
import numpy as np

#create some random values for logits of 10 classes (minibatch size = 1)
logits = np.random.randn(1, 10).astype(np.float32)
target_valid = np.array([3], dtype = np.int32)
target_invalid = np.array([10], dtype = np.int32)

# create cpu variables
logits_cpu = Variable(logits)
target_valid_cpu = Variable(target_valid)
target_invalid_cpu = Variable(target_invalid)

#create gpu variables
logits_gpu = Variable(chainer.cuda.to_gpu(logits))
target_valid_gpu = Variable(chainer.cuda.to_gpu(target_valid))
target_invalid_gpu = Variable(chainer.cuda.to_gpu(target_invalid))

loss = chainer.functions.softmax_cross_entropy(logits_cpu, target_valid_cpu)
print "valid cpu loss is", loss.data
try:
    loss = chainer.functions.softmax_cross_entropy(logits_cpu, target_invalid_cpu)
    print "invalid cpu loss is", loss.data
except IndexError:
    print "cpu index check ok"
    
loss = chainer.functions.softmax_cross_entropy(logits_gpu, target_valid_gpu)
print "valid gpu loss is", loss.data
try:
    loss = chainer.functions.softmax_cross_entropy(logits_gpu, target_invalid_gpu)
    print "invalid gpu loss is", loss.data  # bug here: softmax_cross_entropy should not return a loss, but raise an exception
except IndexError:
    print "gpu index check ok"
```

**Comment by unnonouno** [comment]
Is #302 a same issue?

**Comment by fabiencro** [comment]
I am not sure. The issue in #302 is not very clear to me. It says that if "F.softmax(y) then assert y.max() < y.shape[1]". But I cannot see why we would need y.max() < y.shape[1] when doing F.softmax(y).

But maybe the OP meant : "if F.softmax_cross_entropy(y, t) then t.max() < y.shape[1]". Then it is the same issue.

**Comment by unnonouno** [comment]
We made debug mode that checks range of input variables on softmax_cross_entropy as you say.
https://github.com/pfnet/chainer/blob/v1.7.0/chainer/functions/loss/softmax_cross_entropy.py#L81

And, please note that the debug mode is disabled by default. It is because performance reason as this check requires synchronization between CPU and GPU.

I'll close this issue. Please reopen if you have another problem.
